domain/keyupdater/state/state.go

```diff
@@ -0,0 +1,111 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/canonical/sqlair"
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/core/database"
+	coremachine "github.com/juju/juju/core/machine"
+	"github.com/juju/juju/domain"
+	machineerrors "github.com/juju/juju/domain/machine/errors"
+)
+
+type State struct {
+	*domain.StateBase
+}
+
+// AllPublicKeysQuery returns a state SQL query for fetching the public keys
+// available on a model. This is useful for constructing authorised keys
+// watchers.
+func (s *State) AllPublicKeysQuery() string {
```

Comment:
Where is this going to be used? Does it need to be exported?

Reply:
Yes it is used from the service layer. See the PR linked in the description.

Reply:
strange, I would have thought that raw SQL should not be leaving the state layer

Reply:
Yeah in an ideal world. This is how watchers work for the moment.